### PR TITLE
feat: add msi analysis for TO workflow

### PR DESCRIPTION
**Pull request by khurrammaqbool**
## Description

Add MSI analysis for tumor-only workflow.

[PR description]

#### Added

- MSI analysis for tumor-only workflow

## Documentation

- [X] N/A
- [ ] Updated Balsamic documentation to reflect the changes as needed for this PR.
    - [Document Name]

### Feature Tests

- [ ] N/A
- [ ] Test [Description]
    - [Screenshot]

### Pipeline Integrity Tests

- **Report deliver (generation of the `.hk` file)**
    - [x] N/A
    - [ ] Verified
- **TGA T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **TGA T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **QC Workflow**
    - [x] N/A
    - [ ] Verified
- **PON Workflow**
    - [x] N/A
    - [ ] Verified

## Clinical Genomics Stockholm

<details>
<summary></summary>

### Documentation

- **Atlas documentation**
    - [x] N/A
    - [ ] Updated: [Link]
- **Web portal for Clinical Genomics**
    - [x] N/A
    - [ ] Updated: [Link]

### Panel of Normal specific criteria

- [ ] The PR includes the addition of a new Panel of Normals
- [ ] The samples have been verified to adhere to the sample selection criteria on [Atlas PoN creation instructions for Balsamic](https://atlas.scilifelab.se/infrastructure/BALSAMIC/balsamic_pon/#procedure)

### User Changes

- [x] N/A
- [ ] This PR affects the output files or results.
    - [ ] User feedback is considered unnecessary because [Justification].
    - [ ] Affected users have been included in the development process and given a chance to provide feedback.

### Infrastructure Changes

- **Stored files in Housekeeper**
    - [x] N/A
    - [ ] Updated: [Link]
- **CG (CLI and delivered/uploaded files)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Servers (configuration files on Hasta)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Scout interface**
    - [x] N/A
    - [ ] Updated: [Link]

</details>

## Checklist

> [!IMPORTANT]  
> Ensure that all checkboxes below are ticked before merging.

### For Developers

- **PR Description**
    - [ ] Provided a comprehensive description of the PR.
    - [ ] Linked relevant user stories or issues to the PR.
- **Documentation**
    - [ ] Verified and updated documentation if necessary.
- **Tests**
    - [ ] Described and tested the functionality addressed in the PR.
    - [ ] Ensured integration of the new code with existing workflows.
    - [ ] Confirmed that meaningful unit tests were added for the changes introduced.
    - [ ] Checked that the PR has successfully passed all relevant code smells and coverage checks.
- **Review**
    - [ ] Addressed and resolved all the feedback provided during the code review process.
    - [ ] Obtained final approval from designated reviewers.

### For Reviewers

- **Code**
    - [ ] Code implements the intended features or fixes the reported issue.
    - [ ] Code follows the project's coding standards and style guide.
- **Documentation**
    - [ ] Pipeline changes are well-documented in the CHANGELOG and relevant documentation.
- **Tests**
    - [ ] The author provided a description of their manual testing, including consideration of edge cases and boundary
      conditions where applicable, with satisfactory results.
- **Review**
    - [ ] Confirmed that the developer has addressed all the comments during the code review.

### CODE REVIEW

File: BALSAMIC/commands/config/case.py [hunk 1]
```diff
@@ -167,6 +170,10 @@ def case_config(
             }
         )
 
+    msi_pon_path = {"msi_pon": msi_pon}
+    if msi_pon:
+        references.update(msi_pon_path)
```
Comment (mathiasbio) [inline BALSAMIC/commands/config/case.py:175]:
```suggestion
    if msi_pon:
        references.update({"msi_pon": msi_pon})
```

File: BALSAMIC/snakemake_rules/annotation/msi_tumor_only.rule [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# vim: syntax=python tabstop=4 expandtab
+# coding: utf-8
+# Computation of MSI score.
+
+rule msisensorpro_scan_reference:
```
Comment (mathiasbio) [inline BALSAMIC/snakemake_rules/annotation/msi_tumor_only.rule:5]: Maybe as this rule is identical between tumor normal and tumor only it can be moved into a separate, shared file?

File: BALSAMIC/snakemake_rules/annotation/msi_tumor_only.rule [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    """
+
+rule msisensorpro_msi_tumor_only:
+  input:
+    msi_list=f"{vcf_dir}MSI.somatic.{config['analysis']['case_id']}.msisensorpro.list",
+    bamT= config_model.get_final_bam_name(bam_dir = bam_dir,sample_name=tumor_sample),
+  output:
+    msi_result = f"{vcf_dir}MSI.somatic.{config['analysis']['case_id']}.msisensorpro.msi"
+  benchmark:
+    Path(f"{benchmark_dir}/msisensorpro_msi_tumor_normal_{config['analysis']['case_id']}.tsv").as_posix()
```
Comment (ivadym) [inline BALSAMIC/snakemake_rules/annotation/msi_tumor_only.rule:35]:
```suggestion
    Path(f"{benchmark_dir}/msisensorpro_msi_tumor_only_{config['analysis']['case_id']}.tsv").as_posix()
```

File: CHANGELOG.rst [hunk 4]
```diff
@@ -5,6 +5,7 @@ Added:
 ^^^^^^
 * MSIsensor-pro container https://github.com/Clinical-Genomics/BALSAMIC/pull/1444
 * MSI analysis to the tumor-normal workflow https://github.com/Clinical-Genomics/BALSAMIC/pull/1454
+* MSI analysis to the tumor-only workflow https://github.com/Clinical-Genomics/BALSAMIC/pull/
```
Comment (ivadym) [inline CHANGELOG.rst:8]:
```suggestion
* MSI analysis to the tumor-only workflow https://github.com/Clinical-Genomics/BALSAMIC/pull/1463
```

**Review by ivadym — COMMENTED** [review]
Looks good 👍

**Comment by mathiasbio** [comment]
I forget! We talked a bit about adding validation samples for this feature, like some sample with known high and low MSI which we can use to verify that the feature is working, and continues to keep working for future releases. I don't remember what you said about it, but did we have any such sample to add? 

I understand that actually verifying it for the tumor only validation cases is impossible without the PON, but it would be nice to make sure that we at least have a way to validate it in the validation of release 16 after the PON has been added. 

And for the WGS workflow where a PON exists, do we have a validation sample we could test?

**Comment by mathiasbio** [comment]
Also I'd love it if you could:
- add a bit more in the description, like what the difference is between the tumor normal and the tumor only (that a PON is required)
- add some documentation in Atlas and in balsamic docs about the new MSI PON (same area maybe where we added for CNVkit and for GENS)
- that testing the tumor only workflow for MSI is blocked by the lack of a PON, but maybe give some justification for why this testing shouldn't be necessary (maybe because it basically shares the exact same code as the WGS workflow? I don't know).

File: BALSAMIC/commands/options.py [hunk 5]
```diff
@@ -264,6 +264,14 @@
     show_default=True,
 )
 
+OPTION_MSI_PON = click.option(
+    "--msi-pon",
+    type=click.Path(exists=True, resolve_path=True),
+    required=False,
+    help="Path of MSI PON baseline file",
+)
```
[the remark this answers is not in the document]
Comment (khurrammaqbool) [reply BALSAMIC/commands/options.py:272]: Issue added :https://github.com/Clinical-Genomics/cg/issues/3407

**Comment by khurrammaqbool** [comment]
> I forget! We talked a bit about adding validation samples for this feature, like some sample with known high and low MSI which we can use to verify that the feature is working, and continues to keep working for future releases. I don't remember what you said about it, but did we have any such sample to add?
> 
> I understand that actually verifying it for the tumor only validation cases is impossible without the PON, but it would be nice to make sure that we at least have a way to validate it in the validation of release 16 after the PON has been added.
> 
> And for the WGS workflow where a PON exists, do we have a validation sample we could test?

I added results from MSI analysis using MSI-PON for WGS-TO validation case.
